src/main.zig

```diff
@@ -0,0 +1,411 @@
+    for (numbers, 0..) |n, i| {
+        offset += i;
+        offset += alphabet[n % alphabet.len];
+    }
+    offset %= alphabet.len;
+    offset = (offset + increment) % alphabet.len;
+
+    // Prefix and alphabet.
+    mem.rotate(u8, alphabet, offset);
+    const prefix = alphabet[0];
+    mem.reverse(u8, alphabet);
+
+    // Build the ID.
+    var ret = ArrayList(u8).init(allocator);
+    defer ret.deinit();
+
+    try ret.append(prefix);
+
+    for (numbers, 0..) |n, i| {
+        const x = try toID(allocator, n, alphabet[1..]);
+        defer allocator.free(x);
+        try ret.appendSlice(x);
+
+        if (i < numbers.len - 1) {
+            try ret.append(alphabet[0]);
+            shuffle(alphabet);
+        }
+    }
+
+    // Handle min_length requirements.
+    if (min_length > ret.items.len) {
+        try ret.append(alphabet[0]);
+        while (min_length > ret.items.len) {
+            shuffle(alphabet);
+            const n = @min(min_length - ret.items.len, alphabet.len);
+            try ret.appendSlice(alphabet[0..n]);
+        }
+    }
+
+    var ID = try ret.toOwnedSlice();
+
+    // Handle blocklist.
+    const blocked = try isBlockedID(allocator, blocklist, ID);
+    if (blocked) {
+        allocator.free(ID); // Freeing the old ID string.
+        ID = try encodeNumbers(allocator, numbers, original_alphabet, increment + 1, min_length, blocklist);
+    }
+    return ID;
+}
+
+/// isBlockedID returns true if id collides with the blocklist.
+fn isBlockedID(allocator: mem.Allocator, blocklist: []const []const u8, id: []const u8) !bool {
+    const lower_id = try std.ascii.allocLowerString(allocator, id);
+    defer allocator.free(lower_id);
+
+    for (blocklist) |word| {
+        if (word.len > lower_id.len) {
+            continue;
+        }
+        if (lower_id.len <= 3 or word.len <= 3) {
+            if (mem.eql(u8, id, word)) {
+                return true;
+            }
+        } else if (containsNumber(word)) {
+            if (mem.startsWith(u8, lower_id, word) or mem.endsWith(u8, lower_id, word)) {
+                return true;
+            }
+        } else if (mem.indexOf(u8, lower_id, word)) |_| {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+fn containsNumber(s: []const u8) bool {
+    for (s) |c| {
+        if (std.ascii.isDigit(c)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+/// decode decodes id into numbers using alphabet. Caller owns the memory.
+pub fn decode(
+    allocator: mem.Allocator,
+    to_decode_id: []const u8,
+    decoding_alphabet: []const u8,
+) ![]const u64 {
+    var id = to_decode_id[0..];
+    if (id.len == 0) {
+        return &.{};
+    }
+
+    const alphabet = try allocator.dupe(u8, decoding_alphabet);
+    defer allocator.free(alphabet);
+    shuffle(alphabet);
+
+    // If a character is not in the alphabet, return an empty array.
+    // TODO(lvignoli): here we could return an informative error. Check if fine with specs.
```

Comment:
@4kimov is this well-defined and expected behavior from the specs, or is it not in which case we can return a more informative error to user?

Reply:
The spec just returns an empty array in this case (https://github.com/sqids/sqids-spec/blob/main/src/index.ts#L176). Maybe it's best to do the same for now, and wait for user feedback down the line?